### PR TITLE
Initialising multi gpu training

### PR DESCRIPTION
**Pull request by aaronwtr**
ddp_sharded is not working in PytorchCLI. Opened a bug report in PyTorch Lightning GitHub (see: https://github.com/Lightning-AI/pytorch-lightning/issues/19128 )

This PR contains latest updates that: 
> Runs with batch size 8 on a 80 GB GPU. 
> Added `balance_dataset` from jupyter notebook Will into preprocessing.py. 
> Garbage collection on epoch end 
> Adding a custom collate function to handle proper trimming of target and sequence

### CODE REVIEW

File: config.yaml [hunk 1]
```diff
@@ -13,18 +12,20 @@ trainer:
...
-      mode: min
-      train_time_interval: null #0:60:0
-      every_n_epochs: 1
-      enable_version_counter: true
+    - class_path: lightning.pytorch.callbacks.ModelCheckpoint
+      init_args:
+        save_top_k: -1
+        filename: epoch={epoch}-steps={steps}-train_loss={val/loss:.2f}
+        auto_insert_metric_name: false
+        monitor: train/loss
```
Comment (wconnell) [inline config.yaml:20]: this is inconsistent with the `val/loss` in `filename:`

File: config.yaml [hunk 2]
```diff
@@ -51,7 +52,7 @@ trainer:
   profiler: null
   detect_anomaly: false
   barebones: false
-  plugins: null
+  plugins: ddp_sharded
```
Comment (wconnell) [inline config.yaml]: i think this needs to be a class arg as said in that issue. this can also be hardcoded...

File: eft/data.py [hunk 3]
```diff
@@ -35,6 +47,13 @@ def __init__(self, data_dir: Path = None, batch_size: int = 32, num_workers: int
         self.save_hyperparameters()
         self.fasta_file = str(Path(eft.__file__).parents[1].joinpath('data/hg38.fa'))
 
+    @staticmethod
+    def on_epoch_end():
```
Comment (wconnell) [inline eft/data.py:50]: I dont think this should be req'd, does it help empirically?

File: eft/data.py [hunk 4]
```diff
@@ -22,8 +22,20 @@ def __getitem__(self, ind):
 
         sequence = self.fasta(chr_name, start, end, return_augs=self.return_augs)
 
+        print(f"Sequence dtype: {self.check_tensor_dtype(sequence)}")
```
Comment (wconnell) [inline eft/data.py]: can you just do `str(tensor.dtype`?

**Comment by wconnell** [comment]
I would suggest that you keep working on this as your own branch until we get things sorted out. seems like quite a few of these changes are specific to the troubleshooting / infra that you have. wyt?

File: config.yaml (hunk 2 above)
Comment (awaelchli) [inline config.yaml]: This is not a valid choice in Lightning 2.1 (I assume you have this version since it is written in the comment in this file). 

The distributed strategy is no longer chosen with the "plugins" argument. There is now a separate `strategy` argument for this. For example:

```py
Trainer(strategy="ddp")
```

Furthermore, ddp_sharded was replaced by fsdp, so you probably want to pass
```
strategy: "fsdp" 
```

here, or configure the FSDP strategy further. Tutorial for further reference: https://lightning.ai/docs/pytorch/stable/advanced/model_parallel/fsdp.html#fully-sharded-training


I'm commenting here because I saw the issue you posted over at Lightning. 
Cheers

**Comment by wconnell** [comment]
@aaronwtr FYI: I had issues running `fsdp` strategy with mixed precision. see [here](https://github.com/Lightning-AI/pytorch-lightning/pull/18818) and linked issues. I also had trouble with `deepspeed`. 

I think the fastest solution is to just run `ddp` on a few 40G gpu and we'll get the experiment done in a few hours. Modifying the code to reduce the memory / speed up training won't move the needle much at this point since the training on this smaller dataset is not a huge computational cost.

**Comment by aaronwtr** [comment]
Screenshot 2023-12-27 at 14 06 40

Screenshot 2023-12-27 at 14 06 22

Needed to shrink batch size to 2 to be able to run successfully at 32 bit precision for 20 epochs. See attached loss curves. The tensorboard logs are to big to attach in the PR. I suggest we continue training and calculating metrics + visualising tracks with the config.yaml as currently constructed (barring we get access to more compute).

File: .gitignore
Comment (wconnell) [inline .gitignore:1]: I changed this to except a shared directory for logs. Lets use this to copy runs into that we want to share on master.

**Review by wconnell — APPROVED** [review]

**Comment by wconnell** [comment]
@aaronwtr FYI I moved the `lightning_logs` you pushed into a new shared directory. this is included in the `.gitignore` so we can copy logs we want shared into it.

also, beware I changed the file names a few weeks ago in master to align with the `preprocess.py` outputs
https://github.com/wconnell/enformer-finetune/blob/51c22dde4126ad9dd996058a58c3c6d41f173c7b/eft/data.py#L74-L75
